File: epi_judge_python/even_odd_list_merge.py
from typing import Optional

from list_node import ListNode
from test_framework import generic_test


def even_odd_merge(L: ListNode) -> Optional[ListNode]:
    ev = L
    if L and L.next:
        l2 = od = L.next
    else:
        return L
    while ev.next and od.next:
        ev.next = od.next
        ev = ev.next
        od.next = ev.next
        od = od.next if od.next else od
    ev.next = l2
    od.next = None
    return L


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('even_odd_list_merge.py',
                                       'even_odd_list_merge.tsv',
                                       even_odd_merge))

File: epi_judge_cpp/even_odd_list_merge.cc
#include "list_node.h"
#include "test_framework/generic_test.h"
shared_ptr<ListNode<int>> EvenOddMerge(const shared_ptr<ListNode<int>> &L) {
  if (L && L->next) {
    auto even{L}, odd_head{L->next}, odd{L->next};
    while (even->next) {
      even->next = even->next->next;
      if (odd->next) {
        odd->next = odd->next->next;
        even = even->next;
        odd = odd->next;
      }
    }
    even->next = odd_head;
  }
  return L;
}

int main(int argc, char *argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"L"};
  return GenericTestMain(args, "even_odd_list_merge.cc",
                         "even_odd_list_merge.tsv", &EvenOddMerge,
                         DefaultComparator{}, param_names);
}
